### ISSUE TITLE
local reads ignore nonlocal writes

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#2731]
> I'm writing async tests where a closure tracks call counts via `nonlocal`, then asserts on the count after awaiting. ty infers the variable as its pre-call literal type and treats the assert as unreachable, collapsing everything downstream to `Never`.
> 
> Minimal repro:
> 
> ```python
> """ty bug: nonlocal mutation not tracked → assert narrows to Never.
> 
> ty infers `count` as `Literal[0]` after `await bump()` because it doesn't
> track that the coroutine mutates `count` via `nonlocal`. The subsequent
> `assert count == 1` is seen as always-false, making everything after it
> unreachable (`Never`).
> 
> Run:  uvx ty check ty_nonlocal_bug.py
> """
> 
> 
> async def main() -> None:
>     count = 0
> 
>     async def bump() -> None:
>         nonlocal count
>         count += 1
> 
>     await bump()
> 
>     reveal_type(count)  # ty says: Literal[0]  (should be: int)
>     assert count == 1
>     reveal_type(count)  # ty says: Never       (should be: Literal[1])
> ```
> 
> This is a pretty common pattern in async test suites — tracking side effects through closures and asserting on them. The `Never` cascade makes ty lose all type information for the rest of the test function.
> 
> Just wanted to flag this as a real-world case per the note about revisiting when it comes up in practice. 

 _Originally posted by @chipxsd in [#958](https://github.com/astral-sh/ty/issues/958#issuecomment-3856013529)_

**Comment by carljm** [comment]
Thanks for the report! We should definitely do better here.

For context, pyright and pyrefly both fail on this example in the same way we do. Mypy does not, but only due to its unwillingness to infer literal types. Mypy fails in a similar way on this slightly-modified example:

```py
async def main() -> None:
    count: int | None
    count = 1

    async def bump() -> None:
        nonlocal count
        count = None

    await bump()

    reveal_type(count)  # mypy reveals `int`
    assert count is None
    reveal_type(count)  # mypy thinks this is unreachable
```

So the summary seems to be that this is a failing shared by all Python type checkers today. But I'd like to fix it nonetheless. We may need to prioritize it after areas where we currently fall short of other type checkers, though.

**Comment by oconnor663** [comment]
I think the solution here is similar to machinery I had in https://github.com/astral-sh/ruff/pull/19820? We didn't end up landing that for performance reasons.

**Comment by charliermarsh** [comment]
I played around with this but please feel free to take it @oconnor663 if you want.

**Comment by carljm** [comment]
I don't think we should be prioritizing this issue at all right now; we have more important shortcomings to work on (e.g. any gaps in the conformance suite). This issue isn't even marked for stable release.

**Comment by carljm** [comment]
There are cases that other type checkers do handle better in this space; here's one that came up in #3308:

```py
def outer():
    foo = None
    def inner(set_foo: bool):
        nonlocal foo
        
        if set_foo:
            foo = 1
        else:
            reveal_type(foo)  # `None`, should be `int | None`

    # just illustrating how `foo` can be int in the `else` branch above
    inner(True)
    inner(False)
```

Mypy, zuban and Pycroscope correctly infer `int | None` here; pyright/pyrefly/ty wrongly infer `None`.

I think @oconnor663 is now actually working on this?

**Comment by oconnor663** [comment]
https://github.com/astral-sh/ruff/pull/25387 fixes the first reveal in the first example, though we don't narrow back to `Literal[1]` after the `assert`. It also gives `None | Literal[1]` in Carl's last example.

**Comment by lina-bh** [comment]
I have a case in which ty believes that an Optional, initialised in a callback from an async method, is never assigned to with a non-nil value, and so believes the branch for null always runs. Am I hitting the same bug?

```python
async def discover_aranet4() -> BLEDevice:
    condvar = asyncio.Event()
    aranet4_dev: Optional[BLEDevice] = None

    def det_callback(dev: BLEDevice, _adv: AdvertisementData):
        nonlocal aranet4_dev
        if dev.name is not None and dev.name.startswith("Aranet4"):
            aranet4_dev = dev
            condvar.set()

    async with bleak.BleakScanner(det_callback):
        await condvar.wait()

    if aranet4_dev is None:
        raise Exception("no aranet4")
    return aranet4_dev  # ty says "Code is unreachable"
```

(Ignore that I just realised this code never returns if there is no matching device)

**Comment by carljm** [comment]
@lina-bh Yes that's this issue, and fixed by https://github.com/astral-sh/ruff/pull/25387

**Comment by oconnor663** [comment]
https://github.com/astral-sh/ruff/pull/25387 has landed.
